ZEngine/ZEngine/Core/Container/Array.h

```diff
@@ -0,0 +1,210 @@
+        }
+
+        reference front()
+        {
+            ZENGINE_VALIDATE_ASSERT(m_size > 0, "Index out of range")
+            return m_data[0];
+        }
+
+        const_reference front() const
+        {
+            ZENGINE_VALIDATE_ASSERT(m_size > 0, "Index out of range")
+            return m_data[0];
+        }
+
+        reference back()
+        {
+            ZENGINE_VALIDATE_ASSERT(m_size > 0, "Index out of range")
+            return m_data[m_size - 1];
+        }
+
+        const_reference back() const
+        {
+            ZENGINE_VALIDATE_ASSERT(m_size > 0, "Index out of range")
+            return m_data[m_size - 1];
+        }
+
+        pointer data()
+        {
+            return m_data;
+        }
+
+        const_pointer data() const
+        {
+            return m_data;
+        }
+
+        bool empty() const
+        {
+            return m_size == 0;
+        }
+
+        size_type size() const
+        {
+            return m_size;
+        }
+
+        size_type capacity() const
+        {
+            return m_capacity;
+        }
+
+        void clear()
+        {
+            for (size_type i = 0; i < m_size; ++i)
+            {
+                m_data[i].~T();
+            }
+            m_size = 0;
+        }
+
+        void push(const T& value)
+        {
+            if (m_size == m_capacity)
+            {
+                size_type new_capacity = m_capacity == 0 ? 4 : m_capacity * 2;
+                reserve(new_capacity);
+            }
+            m_data[m_size] = value;
+            ++m_size;
+        }
+
+        void push(T&& value)
+        {
+            if (m_size == m_capacity)
+            {
+                size_type new_capacity = m_capacity == 0 ? 4 : m_capacity * 2;
+                reserve(new_capacity);
+            }
+            m_data[m_size] = std::move(value);
+            ++m_size;
+        }
+
+        reference push_use(const T& value)
+        {
+            if (m_size == m_capacity)
+            {
+                size_type new_capacity = m_capacity == 0 ? 4 : m_capacity * 2;
+                reserve(new_capacity);
+            }
+            m_data[m_size] = value;
+            ++m_size;
+            return back();
+        }
+
+        void pop()
+        {
+            ZENGINE_VALIDATE_ASSERT(m_size > 0, "Index out of range")
+            --m_size;
+            m_data[m_size].~T();
+        }
+
+        ~Array()
+        {
+            clear();
+            m_data     = nullptr;
+            m_capacity = 0;
+        }
+
+        void reserve(size_type new_capacity)
+        {
+            if (new_capacity <= m_capacity)
+            {
+                return;
+            }
+
+            size_t old_alloc_size = m_capacity * sizeof(T);
+            size_t new_alloc_size = new_capacity * sizeof(T);
+
+            m_data                = static_cast<pointer>(m_allocator->Resize(m_data, old_alloc_size, new_alloc_size, alignof(value_type)));
```

Comment:
1. can you create a #define for this allocator->Resize() ,  check the ZEngineDef file
2. alignof(value_type) should be defaulted to DEFAULT_ALIGNMENT is less than it